File: src/decoders/notebook-observation.ts
import type { NotebookObservation, Subject } from "~/models";
import { decodePronoteDate } from "./pronote-date";
import { decodeSubject } from "./subject";

export const decodeNotebookObservation = (observation: any): NotebookObservation => {
  let subject: Subject | undefined;

  if ("L" in observation.matiere.V && observation.matiere.V.N !== "0") {
    subject = decodeSubject(observation.matiere.V);
  }

  return {
    id: observation.N,
    date: decodePronoteDate(observation.date.V),

    opened: observation.estLue,
    shouldParentsJustify: observation.avecARObservation,

    sectionName: observation.L,
    sectionKind: observation.genreObservation,
    sectionID: observation.rubrique.V.N,

    subject,
    reason: observation.commentaire
  };
};

File: library/src/commonMain/kotlin/decoders/notebook-observation.kt
package decoders

import kotlinx.serialization.json.*
import models.NotebookObservation
import models.NotebookObservationKind
import models.Subject

fun decodeNotebookObservation (observation: JsonObject): NotebookObservation {
    var subject: Subject? = null

    if (observation["matiere"]!!.jsonObject["V"]!!.jsonObject.contains("L")
        && observation["matiere"]!!.jsonObject["V"]!!.jsonObject["N"]!!.jsonPrimitive.content != "0")
        subject = decodeSubject(observation["matiere"]!!.jsonObject["V"]!!.jsonObject)

    return NotebookObservation(
        id = observation["N"]!!.jsonPrimitive.content,
        date = decodePronoteDate(observation["date"]!!.jsonObject["V"]!!.jsonPrimitive.content),

        opened = observation["estLue"]!!.jsonPrimitive.boolean,
        shouldParentsJustify = observation["avecARObservation"]!!.jsonPrimitive.boolean,

        name = observation["L"]!!.jsonPrimitive.content,
        kind = NotebookObservationKind.fromInt(observation["genreObservation"]!!.jsonPrimitive.int),
        sectionID = observation["rubrique"]!!.jsonObject["V"]!!.jsonObject["N"]!!.jsonPrimitive.content,

        subject = subject,
        reason = observation["commentaire"]?.jsonPrimitive?.content
    )
}